toolforge-jobs.1
.\" (C) Copyright 2023 Arturo Borrero Gonzalez <aborrero@wikimedia.org>
.\"
.TH TOOLFORGE-JOBS-FRAMEWORK 1 "June 7 2023"
.\" Please adjust this date whenever revising the manpage.
.\"
.SH NAME
toolforge-jobs-framework-cli \- command line interface for the Toolforge Jobs Framework
.SH SYNOPSIS
.B toolforge-jobs [options] {images,run,show,logs,list,delete,flush,load,restart,quota} ...
.SH DESCRIPTION
The \fBtoolforge-jobs\fP command line interface allows you to interact with the \fBToolforge
Jobs Framework\fP.

This framework allows you to manage jobs that run in the \fBWikimedia Toolforge\fP platform.

In general there are 3 kind of jobs:
.TP
.B normal
Jobs that are initiated by the user, and are expected to run until the job's internal execution
normally finishes.
.TP
.B schedule
Jobs that once created by the user, are periodically launched at a given time by the framework.
Similar to a cronjob.
.TP
.B continuous
Jobs that once created by the user are expected to be always up and running (for example, a daemon).

.SH ACTIONS
Top level actions that the command supports:

.TP
.B images
List information on available non-deprecated container image types for the \fBToolforge Jobs Framework\fP.
To be used in the \fBrun\fP command \fB--image\fP parameter.

Occasionally some images are deprecated by the Toolforge admins. In that case existing jobs will continue
working and new jobs can be created, but the images will not be visible on the image listing.

Example:

.nf
$ toolforge-jobs images
+------------+------------------------------------------------------------------------+
| Short name |                          Container image URL                           |
+------------+------------------------------------------------------------------------+
|  bullseye  |    docker-registry.tools.wmflabs.org/toolforge-bullseye-sssd:latest    |
| golang1.11 | docker-registry.tools.wmflabs.org/toolforge-golang111-sssd-base:latest |
|   jdk17    |   docker-registry.tools.wmflabs.org/toolforge-jdk17-sssd-base:latest   |
|  mono6.8   |  docker-registry.tools.wmflabs.org/toolforge-mono68-sssd-base:latest   |
|   node16   |  docker-registry.tools.wmflabs.org/toolforge-node16-sssd-base:latest   |
|  perl5.32  |  docker-registry.tools.wmflabs.org/toolforge-perl532-sssd-base:latest  |
|   php7.4   |   docker-registry.tools.wmflabs.org/toolforge-php74-sssd-base:latest   |
| python3.9  | docker-registry.tools.wmflabs.org/toolforge-python39-sssd-base:latest  |
|  ruby2.1   |  docker-registry.tools.wmflabs.org/toolforge-ruby21-sssd-base:latest   |
|  ruby2.7   |  docker-registry.tools.wmflabs.org/toolforge-ruby27-sssd-base:latest   |
|   tcl8.6   |   docker-registry.tools.wmflabs.org/toolforge-tcl86-sssd-base:latest   |
+------------+------------------------------------------------------------------------+
.fi

.TP
.B run NAME --command COMMAND --image IMAGE --no-filelog --mem MEM --cpu CPU [--schedule SCHEDULE | --continuous | --wait] --retry
Run a new job of your own in Toolforge. Action specific parameters:

.nf
NAME                    New job name, unique identifier. Example: "myjob"
--command COMMAND       The command to run in the job. Example: "./mycommand.sh argone --argtwo"
--image IMAGE           Container image shortname. Check them with the \fBimages\fP action. Example: "bullseye".
--no-filelog            Disable log storage in files in the tool home directory.
-o, --filelog-stdout    Specify the path to store output logs
-e, --filelog-stderr    Specify the path to store error logs
--mem MEM               Request additional memory resource limits for the job.
--cpu CPU               Request additional CPU resource limits for the job
--emails OPT            Specify if you want to receive emails about events for this job. Choices are 'none', 'all', 'onfailure', 'onfinish'. The default is 'none'.

--schedule SCHEDULE     If the job is a schedule, cron time specification. Example: "1 * * * *".
--continuous            Run a continuous job.
--wait                  Run a normal job and wait for completition.
--retry                 Number of times to retry a failed job. This doesn't have any effect when --continuous is set. (range from 0 to 5)
.fi

Some complete examples:

.nf
Running a normal job:
$ toolforge-jobs run myjob --command ./mycommand.sh --image bullseye

Running a normal job and waiting for it to complete:
$ toolforge-jobs run myotherjob --command ./myothercommand.sh --image bullseye --wait

Running a continuous job:
$ toolforge-jobs run myalwaysrunningjob --command ./myendlesscommand.sh --image bullseye --continuous

Running a scheduled job:
$ toolforge-jobs run mycronjob --command ./everyminute.sh --image bullseye --schedule "1 * * * *"

Running a scheduled job with --retry policy set:
$ toolforge-jobs run mycronjob --command ./everyminute.sh --image bullseye --schedule "1 * * * *" --retry

Running a normal job without logs being stored:
$ toolforge-jobs run myjob --command ./mycommand.sh --image bullseye --no-filelog --emails none

Running a normal job with custom log files:
$ toolforge-jobs run myjob --command ./mycommand.sh --image bullseye --emails none -o ./log-file.out -e ./log-file.err

Running a normal job with both output and error logs stored in same file:
$ toolforge-jobs run myjob --command ./mycommand.sh --image bullseye --emails none --filelog-stdout ./log-file.txt --filelog-stderr ./log-file.txt

Running a job with command arguments:
$ toolforge-jobs run myjob --command "./mycommand.sh --witharguments" --image bullseye --emails all

Running a job requesting additional CPU and memory:
$ toolforge-jobs run myjob --command "./heavycommand.sh" --image bullseye --mem 1Gi --cpu 10
.fi

.TP
.B show NAME
Show details of a job of your own in Toolforge.

Example:

.nf
$ toolforge-jobs show myscheduledjob
+------------+-----------------------------------------------------------------+
| Job name:  | myscheduledjob                                                  |
+------------+-----------------------------------------------------------------+
| Command:   | ./read-dumps.sh myargument                                      |
+------------+-----------------------------------------------------------------+
| Job type:  | schedule: * * * * *                                             |
+------------+-----------------------------------------------------------------+
| Container: | bullseye                                                        |
+------------+-----------------------------------------------------------------+
| File log:  | yes                                                             |
+------------+-----------------------------------------------------------------+
| Output log:| /data/project/my-user/myscheduledjob.out                        |
+------------+-----------------------------------------------------------------+
| Error log: | /data/project/my-user/myscheduledjob.err                        |
+------------+-----------------------------------------------------------------+
| Emails:    | none                                                            |
+------------+-----------------------------------------------------------------+
| Resources: | mem: 10Mi, cpu: 100                                             |
+------------+-----------------------------------------------------------------+
| Retry:     | yes: 3 time(s)                                                  |
+------------+-----------------------------------------------------------------+
| Status:    | Last schedule time: 2021-06-30T10:26:00Z                        |
+------------+-----------------------------------------------------------------+
| Hints:     | Last run at 2021-06-30T10:26:08Z. Pod in 'Pending' phase. State |
|            | 'waiting' for reason 'ContainerCreating'.                       |
+------------+-----------------------------------------------------------------+
.fi

.TP
.B logs [-f|--follow] [-l|--lat LAST] NAME
Display log output from a currently running job.

.TP
.B list [-o|--output {normal,long}]
List all running jobs of your own in Toolforge.

The \fB-o\fP (or \fB--output\fP) parameter indicates how much detail is displayed.

Example, short listing:

.nf
$ toolforge-jobs list
Job name:       Job type:            Status:
--------------  -------------------  ---------------------------
myscheduledjob  schedule: * * * * *  Last schedule time: 2021-06-30T10:26:00Z
alwaysrunning   continuous           Running
myjob           normal               Completed
.fi

Example, long listing:

.nf
$ toolforge-jobs list -o long
Job name:       Command:                 Job type:            Container:           File log:  Output log:                               Error log:                                Emails:   Resources:  Retry:           Status:
--------------  -----------------------  -------------------  -------------------  ---------  ----------------------------------------  ----------------------------------------  -------   ----------  --------------   ----------------------------------------
myscheduledjob  ./read-dumps.sh          schedule: * * * * *  bullseye              yes       /data/project/my-user/myscheduledjob.out  /data/project/my-user/myscheduledjob.err  none      default     yes: 5 time(s)   Last schedule time: 2021-06-30T10:26:00Z
alwaysrunning   ./myendlesscommand.sh    continuous           bullseye              no        /dev/null                                 /dev/null                                 all       default     no               Running
myjob           ./mycommand.sh --debug   normal               buster (deprecated)   yes       /data/project/my-user/custom.out          /data/project/my-user/custom.err          onfinish  default     yes: 1 time(s)   Completed
.fi

.TP
.B delete NAME
Delete a running job of your own in Toolforge.
.TP
.B flush
Delete all running jobs of your own in Toolforge.
.TP
.B load FILE
Flush all jobs (similar to \fBflush\fP action) and read a YAML file with job specifications to be
loaded and run all at once.

Loading new jobs will stop if failures are found.

The file format mirrors arguments to the \fBrun\fP action.

Example YAML file:

.nf
---
# a cronjob
- name: everyminute
  command: ./myothercommand.py -v
  image: bullseye
  no-filelog: true
  schedule: "* * * * *"
  retry: 5
  emails: onfailure
# a continuous job
- image: bullseye
  name: endlessjob
  command: ./dumps-daemon.py --endless

debian/changelog
toolforge-jobs-framework-cli (14) buster; urgency=medium

  * [f9ec42b] cli: Use new more structured URL routes
  * [0d30629] cli: run: Use JSON in POST body

 -- Taavi Väänänen <hi@taavi.wtf>  Mon, 09 Oct 2023 09:51:02 +0300

toolforge-jobs-framework-cli (13) buster; urgency=medium

  [ Arturo Borrero Gonzalez ]
  * [90805dc] cli: remove --cert and --key options

  [ Taavi Väänänen ]
  * [f487c01] cli: remove 'containers' option, replaced by 'images'
  * [8d23473] cli: Increase timeout (and make it configurable)
  * [1a4c05c] cli: Use toolforge-weld config system
  * [dadbc94] Add support for querying logs

 -- Taavi Väänänen <hi@taavi.wtf>  Thu, 05 Oct 2023 12:01:35 +0300

toolforge-jobs-framework-cli (12) buster; urgency=medium

  [ Arturo Borrero Gonzalez ]
  * [cee44da] d/source/options: prevent .tox directory to be present in
    the source package

  [ Taavi Väänänen ]
  * [25930c2] errors: Print context of parent if necessary
  * [9c0c85e] cli: Print which job failed to load

  [ David Caro ]
  * [d08f7fd] k8s: allow loading relative kubeconfig certs

  [ Siddharth VP ]
  * [7953469] Fix memory flag normalization while comparing job config
    with API output

  [ Taavi Väänänen ]
  * [010d465] cli: Hide admin-only options from --help output
  * [0333d78] cli: use debug env variable from toolforge-cli
  * [c8f746a] cli: use 'toolforge jobs' as help command when used in the CLI
  * [ea55b96] cli: list: add -o name
  * [bb5a955] Add tab completion

  [ Arturo Borrero Gonzalez ]
  * [dfb0e42] tests/cmd-checklist: drop unused stuff
  * [0c8d553] jobs-framework-cli: drop custom address support
  * [442e1d7] tests/cmd-checklist: relocate test-cmd.sh file to lima-kilo
  * [c8ac035] jobs-framework-cli: use toolforge-weld

  [ Taavi Väänänen ]
  * [17a1adf] Add quota subcommand
  * [ca44f12] d/: add gbp.conf to set `gbp dch` options
  * [e7b0ea8] d/: configure py3dist-overrides for -weld

 -- Taavi Väänänen <hi@taavi.wtf>  Mon, 12 Jun 2023 11:03:23 +0300

toolforge-jobs-framework-cli (11) buster; urgency=medium

  [ Taavi Väänänen ]
  * [b606fd3] cmd-checklist-runner: format with black
  * [c26bae5] tox: pytest: install in editable mode
  * [c2ef1ac] loader: add handling for the retry default value

  [ Arturo Borrero Gonzalez ]
  * [6af7a85] cli: drop unused _flush_and_wait() function
  * [10fc50c] tests: add support for new api-gateway setup
  * [a8e5383] cli: introduce support for lima-kilo latest isolation setup

  [ Taavi Väänänen ]
  * [0be49cd] cli: list: use a more general --output flag
  * [22f878c] tox: remove skipdist
  * [2c8d855] conf: Raise proper errors instead of quitting directly
  * [87b8af7] conf: rename as api
  * [efff4c7] api: add wrappers for http requests
  * [4dc1bee] Overhaul error handling

 -- Arturo Borrero Gonzalez <aborrero@wikimedia.org>  Tue, 11 Apr 2023 16:21:46 +0200

toolforge-jobs-framework-cli (10) buster; urgency=medium

  [ Arturo Borrero Gonzalez ]
  * [4012f7c] d/py3dist-overrides: override forcediphttpsadapter

 -- Taavi Väänänen <hi@taavi.wtf>  Tue, 24 Jan 2023 13:53:50 +0200

toolforge-jobs-framework-cli (9) buster; urgency=medium

  [ Raymond Ndibe ]
  * [812a031] jobs-framework-cli: minor improvements to test file

  [ Arturo Borrero Gonzalez ]
  * [9b0740e] .gitignore: ignore autogenerated egg-info directory

  [ Raymond Nbide ]
  * [2547930] add --filelog-stdout and --filelog-stderr args to cli
  * [ac8b96d] add --retry to cli

  [ Taavi Väänänen ]
  * [2f93477] test_loader: drop parameters no longer emitted by the api

  [ Arturo Borrero Gonzalez ]
  * [763c493] gitignore: ignore build/ and dist/
  * [441cf41] jobs-framework-cli: adapt to the lima-kilo project
  * [c29d659] setup.py: add forcediphttpsadapter as dependency
  * [2328cbc] tests: relocate pytest tests

  [ Taavi Väänänen ]
  * [d203c6b] Adjust for image name changes

  [ Arturo Borrero Gonzalez ]
  * [6b6c97e] tests/: wait for jobs to complete before checking resulsts
  * [ab42014] loader: make it understand filelog, filelog_stdout, filelog_stderr
  * [e5ae611] cli: warn about unknown keys in the load YAML file

 -- Taavi Väänänen <hi@taavi.wtf>  Tue, 24 Jan 2023 11:57:22 +0200

toolforge-jobs-framework-cli (8) buster; urgency=medium

  [ JJMC89 ]
  * [897c4af] fix typo in error message when attempting to show a nonexistent job

  [ Klein Muçi ]
  * [b47b586] Fix typo

  [ Taavi Väänänen ]
  * [124dcdd] Set up as a proper Python package
  * [6e9bf40] load: Support loading a single job only
  * [19d57ce] Add restart subcommand
  * [ae5ce81] load: Don't change already running jobs
  * [e7a64ce] d/rules: Don't run pytest when building the package

 -- Taavi Väänänen <hi@taavi.wtf>  Tue, 18 Oct 2022 14:42:35 +0300

toolforge-jobs-framework-cli (7) buster; urgency=medium

  * [8677d69] images: update API endpoint URL
  * [38b6d5f] jobs-framework-cli: only print timestamp and file in debug mode

 -- Arturo Borrero Gonzalez <arturo@debian.org>  Tue, 05 Apr 2022 09:15:47 +0200

toolforge-jobs-framework-cli (6) buster; urgency=medium

  * [7347e84] tests/: refresh the checklist runner script
  * [adacda7] jobs-framework-cli: add run_ci_locally.sh script

toolforge-jobs.py
#!/usr/bin/env python3

# (C) 2021 by Arturo Borrero Gonzalez <aborrero@wikimedia.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# Some funcionts of this code were copy-pasted from the tools-webservice package.
# Copyright on that TBD.
#
# This program is the command line interface part of the Toolforge Jobs Framework.
#

from tabulate import tabulate
import textwrap
import requests
import argparse
import getpass
import urllib3
import logging
import socket
import time
import json
import yaml
import sys
import os


# TODO: disable this for now, review later
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

# for --wait: 5 minutes timeout, check every 5 seconds
WAIT_TIMEOUT = 60 * 5
WAIT_SLEEP = 5


class Conf:
    """
    Class that represents the configuration for this CLI session
    """

    JOB_TABULATION_HEADERS_SHORT = {
        "name": "Job name:",
        "type": "Job type:",
        "status_short": "Status:",
    }

    JOB_TABULATION_HEADERS_LONG = {
        "name": "Job name:",
        "cmd": "Command:",
        "type": "Job type:",
        "image": "Image:",
        "filelog": "File log:",
        "emails": "Emails:",
        "resources": "Resources:",
        "status_short": "Status:",
        "status_long": "Hints:",
    }

    IMAGES_TABULATION_HEADERS = {
        "shortname": "Short name",
        "image": "Container image URL",
    }

    def __init__(self, cfg_file: str):
        """Constructor"""

        try:
            with open(cfg_file) as f:
                cfg = yaml.safe_load(f.read())
        except Exception as e:
            logging.error(
                f"couldn't read config file '{cfg_file}': {e}. Contact a Toolforge admin."
            )
            sys.exit(1)

        try:
            self.api_url = cfg.get("api_url")
        except KeyError as e:
            logging.error(
                f"missing key '{str(e)}' in config file '{cfg_file}'. Contact a Toolforge admin."
            )
            sys.exit(1)

        kubeconfig = cfg.get("kubeconfig", "~/.kube/config")
        customhdr = cfg.get("customhdr", None)
        customaddr = cfg.get("customaddr", None)
        customfqdn = cfg.get("customfqdn", None)
        self.kubeconfigfile = os.path.expanduser(kubeconfig)

        try:
            with open(self.kubeconfigfile) as f:
                self.k8sconf = yaml.safe_load(f.read())
        except Exception as e:
            logging.error(
                f"couldn't read kubeconfig file '{self.kubeconfigfile}': {e}. "
                "Contact a Toolforge admin."
            )
            sys.exit(1)

        logging.debug(f"loaded kubeconfig file '{self.kubeconfigfile}'")

        self.session = requests.Session()
        try:
            self.context = self._find_cfg_obj("contexts", self.k8sconf["current-context"])
            self.cluster = self._find_cfg_obj("clusters", self.context["cluster"])
            self.server = self.cluster["server"]
            self.namespace = self.context["namespace"]
            self.user = self._find_cfg_obj("users", self.context["user"])
            self.session.cert = (self.user["client-certificate"], self.user["client-key"])
        except KeyError as e:
            logging.error(
                "couldn't build session configuration from file "
                f"'{self.kubeconfigfile}': missing key {e}. Contact a Toolforge admin."
            )
            sys.exit(1)
        except Exception as e:
            logging.error(
                "couldn't build session configuration from file "
                f"'{self.kubeconfigfile}': {e}. Contact a Toolforge admin."
            )
            sys.exit(1)

        self._configure_user_agent()

        if customhdr is not None:
            self.session.headers.update(customhdr)

        # don't verify server-side TLS for now
        self.session.verify = False

        if customaddr is not None and customfqdn is not None:
            from forcediphttpsadapter.adapters import ForcedIPHTTPSAdapter

            self.session.mount(f"https://{customfqdn}", ForcedIPHTTPSAdapter(dest_ip=customaddr))

    def _configure_user_agent(self):
        """Configure User-Agent header."""
        host = socket.gethostname()
        pyrequest_ua = requests.utils.default_user_agent()
        ua_str = f"jobs-framework-cli {self.namespace}@{host} {pyrequest_ua}"
        self.session.headers.update({"User-Agent": ua_str})

    def _find_cfg_obj(self, kind, name):
        """Lookup a named object in our config."""
        for obj in self.k8sconf[kind]:
            if obj["name"] == name:
                return obj[kind[:-1]]
        raise KeyError(f"key '{name}' not found in '{kind}' section of config")


def parse_args():
    description = "Toolforge Jobs Framework, command line interface"
    parser = argparse.ArgumentParser(description=description)

    parser.add_argument("--debug", action="store_true", help="activate debug mode")
    parser.add_argument(
        "--cfg",
        default="/etc/toolforge-jobs-framework-cli.cfg",
        help="YAML config for the CLI. Defaults to '%(default)s'. "
        "Only useful for Toolforge admins.",
    )

    subparser = parser.add_subparsers(
        help="possible operations (pass -h to know usage of each)",
        dest="operation",
        required=True,
    )

    # TODO: remove this after a few months
    subparser.add_parser(
        "containers",
        help="Kept for compatibility reasons, use `images` instead.",
    )
    subparser.add_parser(
        "images",
        help="list information on available container image types for Toolforge jobs",
    )

    runparser = subparser.add_parser(
        "run",
        help="run a new job of your own in Toolforge",
    )

    runparser.add_argument("name", help="new job name")
    runparser.add_argument(
        "--command", required=True, help="full path of command to run in this job"
    )
    runparser.add_argument(
        "--image", required=True, help="image shortname (check them with `images`)"
    )
    runparser.add_argument(
        "--no-filelog",
        required=False,
        action="store_true",
        help="don't store job stdout in `jobname`.out and stderr in `jobname`.err files in the "
        "user home directory",
    )
    runparser.add_argument(
        "--mem",
        required=False,
        help="specify additional memory limit required for this job",
    )
    runparser.add_argument(
        "--cpu",
        required=False,
        help="specify additional CPU limit required for this job",
    )
    runparser.add_argument(
        "--emails",
        required=False,
        choices=["none", "all", "onfinish", "onfailure"],
        default="none",
        help="specify if the system should email notifications about this job. "
        "Defaults to '%(default)s'.",
    )

    runparser_exclusive_group = runparser.add_mutually_exclusive_group()
    runparser_exclusive_group.add_argument(
        "--schedule",
        required=False,
        help="run a job with a cron-like schedule (example '1 * * * *')",
    )
    runparser_exclusive_group.add_argument(
        "--continuous", required=False, action="store_true", help="run a continuous job"
    )
    runparser_exclusive_group.add_argument(
        "--wait",
        required=False,
        action="store_true",
        help=f"run a job and wait for completition. Timeout is {WAIT_TIMEOUT} seconds.",
    )

    showparser = subparser.add_parser(
        "show",
        help="show details of a job of your own in Toolforge",
    )
    showparser.add_argument("name", help="job name")

    listparser = subparser.add_parser(
        "list",
        help="list all running jobs of your own in Toolforge",
    )
    listparser.add_argument(
        "-l",
        "--long",
        required=False,
        action="store_true",
        help="show long table with full details about each job",
    )

    deleteparser = subparser.add_parser(
        "delete",
        help="delete a running job of your own in Toolforge",
    )
    deleteparser.add_argument("name", help="job name")

    subparser.add_parser(
        "flush",
        help="delete all running jobs of your own in Toolforge",
    )

    loadparser = subparser.add_parser(
        "load",
        help="flush all jobs and load a YAML file with job definitions and run them",
    )
    loadparser.add_argument("file", help="path to YAML file to load")

    return parser.parse_args()


def op_images(conf: Conf):
    try:
        response = conf.session.get(conf.api_url + "/images/")
    except Exception as e:
        logging.error(f"couldn't contact the API endpoint. Contact a Toolforge admin: {e}")
        sys.exit(1)

    if response.status_code != 200:
        logging.error(f"unable to fetch information. Contact a Toolforge admin: {response.text}")
        sys.exit(1)

    try:
        images = json.loads(response.text)
    except Exception as e:
        logging.error(f"couldn't parse information from the API. Contact a Toolforge admin: {e}")
        sys.exit(1)

    try:
        output = tabulate(images, headers=conf.IMAGES_TABULATION_HEADERS, tablefmt="pretty")
    except Exception as e:
        logging.error(f"couldn't format information from the API. Contact a Toolforge admin: {e}")
        sys.exit(1)

    print(output)


def job_prepare_for_output(conf: Conf, job, long_listing=False, supress_hints=True):
    schedule = job.get("schedule", None)
    cont = job.get("continuous", None)
    if schedule is not None:
        job["type"] = f"schedule: {schedule}"
        job.pop("schedule", None)
    elif cont is not None:
        job["type"] = "continuous"
        job.pop("continuous", None)
    else:
        job["type"] = "normal"

    filelog = job.get("filelog", "false")
    if filelog == "True":
        job["filelog"] = "yes"
    else:
        job["filelog"] = "no"

    mem = job.pop("memory", "default")
    cpu = job.pop("cpu", "default")
    if mem == "default" and cpu == "default":
        job["resources"] = "default"
    else:
        job["resources"] = f"mem: {mem}, cpu: {cpu}"

    if supress_hints:
        if job.get("status_long", None) is not None:
            job.pop("status_long", None)
    else:
        job["status_long"] = textwrap.fill(job.get("status_long", "Unknown"))

    if long_listing:
        headers = conf.JOB_TABULATION_HEADERS_LONG
    else:
        headers = conf.JOB_TABULATION_HEADERS_SHORT

    # not interested in these fields ATM
    for key in job.copy():
        if key not in headers:
            logging.debug(f"supressing job API field '{key}' before output")
            job.pop(key)

    # normalize key names for easier printing
    for key in headers:
        if key == "status_long" and supress_hints:
            continue

        oldkey = key
        newkey = headers[key]
        job[newkey] = job.pop(oldkey, "Unknown")


def _list_jobs(conf: Conf):
    try:
        response = conf.session.get(conf.api_url + "/list/")
    except Exception as e:
        logging.error(f"couldn't contact the API endpoint. Contact a Toolforge admin: {e}")
        sys.exit(1)

    if response.status_code != 200:
        logging.error(f"unable to fetch information. Contact a Toolforge admin: {response.text}")
        sys.exit(1)

    try:
        list = json.loads(response.text)
    except Exception as e:
        logging.error(f"couldn't parse information from the API. Contact a Toolforge admin: {e}")
        sys.exit(1)

    return list


def op_list(conf: Conf, long_listing: bool):
    list = _list_jobs(conf)

    if len(list) == 0:
        logging.debug("no jobs to be listed")
        return

    try:
        for job in list:
            logging.debug(f"job information from the API: {job}")
            job_prepare_for_output(conf, job, supress_hints=True, long_listing=long_listing)

        if long_listing:
            headers = conf.JOB_TABULATION_HEADERS_LONG
        else:
            headers = conf.JOB_TABULATION_HEADERS_SHORT

        output = tabulate(list, headers=headers, tablefmt="pretty")
    except Exception as e:
        logging.error(f"couldn't format information from the API. Contact a Toolforge admin: {e}")
        sys.exit(1)

    print(output)


def _wait_for_job(conf: Conf, name: str):
    curtime = starttime = time.time()
    while curtime - starttime < WAIT_TIMEOUT:
        time.sleep(WAIT_SLEEP)
        curtime = time.time()

        job = _show_job(conf, name, missing_ok=True)
        if job is None:
            logging.info(f"job '{name}' completed (and already deleted)")
            return

        if job["status_short"] == "Completed":
            logging.info(f"job '{name}' completed")
            return

        if job["status_short"] == "Failed":
            logging.error(f"job '{name}' failed:")
            op_show(conf, name)
            sys.exit(1)

    logging.error(f"timed out {WAIT_TIMEOUT} seconds waiting for job '{name}' to complete:")
    op_show(conf, name)
    sys.exit(1)


def op_run(
    conf: Conf,
    name,
    command,
    schedule,
    continuous,
    image,
    wait,
    no_filelog: bool,
    mem: str,
    cpu: str,
    emails: str,
):
    payload = {"name": name, "imagename": image, "cmd": command, "emails": emails}

    if continuous:
        payload["continuous"] = "true"
    elif schedule:
        payload["schedule"] = schedule

    if not no_filelog:
        # the default is to request the filelog
        payload["filelog"] = "true"

    if mem:
        payload["memory"] = mem

    if cpu:
        payload["cpu"] = cpu

    logging.debug(f"payload: {payload}")

    try:
        response = conf.session.post(conf.api_url + "/run/", data=payload)
    except Exception as e:
        logging.error(f"couldn't contact the API endpoint. Contact a Toolforge admin: {e}")
        sys.exit(1)

    if response.status_code == 409:
        logging.error(f"a job with the same name '{name}' exists already")
        sys.exit(1)

    if response.status_code >= 300:
        logging.error(f"unable to create job: {response.text.strip()}")
        sys.exit(1)

    logging.debug("job was created")

    if wait:
        _wait_for_job(conf, name)


def _show_job(conf: Conf, name: str, missing_ok: bool):
    try:
        response = conf.session.get(conf.api_url + f"/show/{name}")
    except Exception as e:
        logging.error(f"couldn't contact the API endpoint. Contact a Toolforge admin: {e}")
        sys.exit(1)

    if response.status_code == 404:
        if missing_ok:
            return None  # the job doesn't exist, but that's ok!

        logging.error(f"job '{name}' does not exist")
        sys.exit(1)

    try:
        job = json.loads(response.text)
    except Exception as e:
        logging.error(f"couldn't parse information from the API. Contact a Toolforge admin: {e}")
        sys.exit(1)

    logging.debug(f"job information from the API: {job}")
    return job


def op_show(conf: Conf, name):
    job = _show_job(conf, name, missing_ok=False)
    job_prepare_for_output(conf, job, supress_hints=False, long_listing=True)

    # change table direction
    kvlist = []
    for key in job:
        kvlist.append([key, job[key]])

    try:
        output = tabulate(kvlist, tablefmt="grid")
    except Exception as e:
        logging.error(f"couldn't format information from the API. Contact a Toolforge admin: {e}")
        sys.exit(1)

    print(output)


def op_delete(conf: Conf, name):
    try:
        conf.session.delete(conf.api_url + f"/delete/{name}")
    except Exception as e:
        logging.error(f"couldn't contact the API endpoint. Contact a Toolforge admin: {e}")
        sys.exit(1)

    logging.debug("job was deleted (if it existed anyway, we didn't check)")


def op_flush(conf: Conf):
    try:
        conf.session.delete(conf.api_url + "/flush/")
    except Exception as e:
        logging.error(f"couldn't contact the API endpoint. Contact a Toolforge admin: {e}")
        sys.exit(1)

    logging.debug("all jobs were flushed (if any existed anyway, we didn't check)")


def _flush_and_wait(conf: Conf):
    op_flush(conf)

    curtime = starttime = time.time()
    while curtime - starttime < WAIT_TIMEOUT:
        logging.debug(f"waiting for jobs list to be empty, sleeping {WAIT_SLEEP} seconds")
        time.sleep(WAIT_SLEEP)
        curtime = time.time()

        list = _list_jobs(conf)

        if len(list) == 0:
            # ok!
            return

    logging.error("could not load new jobs")
    logging.error(f"timed out ({WAIT_TIMEOUT} seconds) waiting for previous jobs to be flushed")
    sys.exit(1)


def _load_job(conf: Conf, job: dict, n: int):
    # these are mandatory
    try:
        name = job["name"]
        command = job["command"]
        image = job["image"]
    except KeyError as e:
        logging.error(f"Unable to load job number {n}. Missing configuration parameter {str(e)}")
        sys.exit(1)

    # these are optional
    schedule = job.get("schedule", None)
    continuous = job.get("continuous", False)
    no_filelog = job.get("no-filelog", False)
    mem = job.get("mem", None)
    cpu = job.get("cpu", None)
    emails = job.get("emails", "none")

    if not schedule and not continuous:
        wait = job.get("wait", False)
    else:
        wait = False

    op_run(
        conf=conf,
        name=name,
        command=command,
        schedule=schedule,
        continuous=continuous,
        image=image,
        wait=wait,
        no_filelog=no_filelog,
        mem=mem,
        cpu=cpu,
        emails=emails,
    )


def op_load(conf: Conf, file: str):
    try:
        with open(file) as f:
            jobslist = yaml.safe_load(f.read())
    except Exception as e:
        logging.error(f"couldn't read YAML file '{file}': {e}")
        sys.exit(1)

    logging.debug(f"loaded content from YAML file '{file}':")
    logging.debug(f"{jobslist}")

    # before loading new jobs, flush and wait for them to go away
    _flush_and_wait(conf)
    logging.debug("jobs list is confirmed to be empty, now loading new jobs")

    for n, job in enumerate(jobslist, start=1):
        _load_job(conf, job, n)


def main():
    args = parse_args()

    logging_format = "%(levelname)s: %(message)s"
    if args.debug:
        logging_level = logging.DEBUG
        logging_format = f"[%(asctime)s] [%(filename)s] {logging_format}"
    else:
        logging_level = logging.INFO

    logging.addLevelName(
        logging.WARNING, "\033[1;33m%s\033[1;0m" % logging.getLevelName(logging.WARNING)
    )
    logging.addLevelName(
        logging.ERROR, "\033[1;31m%s\033[1;0m" % logging.getLevelName(logging.ERROR)
    )
    logging.basicConfig(
        format=logging_format, level=logging_level, stream=sys.stdout, datefmt="%Y-%m-%d %H:%M:%S"
    )

    user = getpass.getuser()
    if not user.startswith("tools.") and not user.startswith("toolsbeta."):
        logging.warning(
            "not running as the tool account? Likely to fail. Perhaps you forgot `become <tool>`?"
        )

    conf = Conf(args.cfg)
    logging.debug("session configuration generated correctly")

    if args.operation == "images":
        op_images(conf)
    elif args.operation == "containers":
        # TODO: remove this after a few months
        logging.warning("the `containers` action is deprecated. Use `images` instead.")
        op_images(conf)
    elif args.operation == "run":
        op_run(
            conf,
            args.name,
            args.command,
            args.schedule,
            args.continuous,
            args.image,
            args.wait,
            args.no_filelog,
            args.mem,
            args.cpu,
            args.emails,
        )
    elif args.operation == "show":
        op_show(conf, args.name)
    elif args.operation == "delete":
        op_delete(conf, args.name)
    elif args.operation == "list":
        op_list(conf, args.long)
    elif args.operation == "flush":
        op_flush(conf)
    elif args.operation == "load":
        op_load(conf, args.file)

    logging.debug("-- end of operations")


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(130)
